Update Testrunner documentation (#583)
SUSE/skuba: ci/infra/testrunner/README.md

# Testrunner

## Contents

- [Summary](#summary)
- [Configuration](#configuration-parameters)
  - [Work Environment](#work-environment)
  - [Platform](#platform)
    - [Terraform](#terraform)
    - [Openstack](#openstack)
    - [VMware](#vmware)
  - [Skuba](#skuba)
  - [Log](#log)
  - [Test](#test)
- [Environment setup](#environment-setup)
  - [Local setup](#local-setup)
  - [CI setup](#ci-setup)
- [Usage](#usage)
  - [General CLI options](#general-cli-options)
  - [Provision](#provision)
  - [Node commands](#node-commands)
    - [Node Upgrade](#node-upgrade)
  - [Ssh](#ssh)
  - [Test](#ssh)
- [Examples](#examples)
  - [Create K8s Cluster](#create-k8s-cluster)
  - [Collect logs](#collect-logs)

## Summary

> Testrunner is a CLI tool for setting up a running e2e tests, making transparent the mechanism used for providing the test infrastructure. It can used as a stand-alone tool, running either locally o as part of a ci pipeline. It provides commands for provisioning the infrastructure, deploying a k8s cluster using `skuba` and running tests. It also provides [a library](tests/README.md) for developing `pytest`-based tests.

Reviewer: s/locally o as/locally or as/

Author: I don't really understand the suggestion. Sorry, regular expressions are not one of my strengths :sweat_smile:

Author: I finally got it :sweat_smile:

Reviewer: Sorry. 🙈 In some teams, there is a bot that is using this syntax and updating the pr accordingly. 
I will adapt next time.

> Testrunner is a CLI tool for setting up a running e2e tests, making transparent the mechanism used for providing the test infrastructure. It can used as a stand-alone tool, running either locally o as part of a ci pipeline. It provides commands for provisioning the infrastructure, deploying a k8s cluster using `skuba` and running tests. It also provides [a library](tests/README.md) for developing `pytest`-based tests.

Reviewer 2: ```suggestion
Testrunner is a CLI tool for setting up an environment for running e2e tests, making transparent the mechanism used for providing the test infrastructure. It can used as a stand-alone tool, running either locally or as part of a ci pipeline. It provides commands for provisioning the infrastructure, deploying a k8s cluster using `skuba` and running tests. It also provides [a library](tests/README.md) for developing `pytest`-based tests.
```

## Configuration parameters

Testrunner provides configuration by means of:
 
- A yaml configuration file (defaults to `vars.yaml` in current directory)

> - Environment variables that override the configuration. Every configuration option of the form `<section>.<variable>' can be subtituted by an environment variable `SECTION_VARIABLE. Notice that some variables are defined in the "root" section (e.g `workspace`). For example `skuba.binpath` is override by `SKUBA_BINPATH` and `workspace` by `WORKSPACE`

Reviewer 2: ```suggestion
- Environment variables that override the configuration. Every configuration option of the form `<section>.<variable>' can be subtituted by an environment variable `SECTION_VARIABLE. Notice that some variables are defined in the "root" section (e.g `workspace`). For example `skuba.binpath` is overriden by `SKUBA_BINPATH` and `workspace` by `WORKSPACE`
```

> - CLI options which override some configuration parameters

Author: Good point. I maybe it would be hard to maintain? I'll make a general commend that when a cli option overrides a configuration option?

> - CLI options which override some configuration parameters

Author: Good point. I think this should be documented in the help for the corresponding options, to avoid having to document in two places.

The following sections document the configuration options. The CLI arguments are described in the [Usage section](#usage).

### Work Environment

This section configures the working environment and is generally specific of each user of CI job-

- workspace: path to the testrunner's working directory 
- username: userid for the user. It's optional. If `platform.stack_name` is not set, `username` is used. 

```
workspace "/path/to/your/workspace" 

> username: "mysuserid"

Reviewer 3: `myuserid` ? or maybe `USERNAME` so it is extra obvious

Author: USERNAME is an environment variable. I wanted to avoid confusion.

```

### Platform

There are some arguments that are currently at the top level of the configuration, but are actually related to the platform:

- log_dir: path to the directory where platform logs are collected. Defaults to `<workspace>/testrunner_logs` 
- nodeuser: the userid used to login into the platform nodes. Optional. 
- ssh_option: specifies the location of the key used to access nodes. Can have two possible values:
  - "id_shared": uses the key located at `<skuba src directory>/ci/infra/id_shared` (default)
  - "id_rsa": uses the user's key located at `$HOME/.ssh/id_rsa`

#### Terraform

General setting for terraform-based platforms such as [Openstack](#openstack) and [VMware](#vmware). 

* internal_net: name of the network used when provisioning the platform. Defaults to `stack_name`

> * mirror: URL for the repository mirrors to be used when setting the skuba nodes, replacing the URL of the repositories defined in terraform. Used, for instance to switch to development repositories or internal repositories when running in the ci pipeline.

Reviewer 2: ```suggestion
* mirror: URL for the repository mirrors to be used when setting up the skuba nodes, replacing the URL of the repositories defined in terraform. Used, for instance to switch to development repositories or internal repositories when running in the ci pipeline.
```

> * mirror: URL for the repository mirrors to be used when setting up the skuba nodes, replacing the URL of the repositories defined in terraform. Used, for instance to switch to development repositories or internal repositories when running in the CI pipeline.

Reviewer 4: ```suggestion
* mirror: URL for the repository mirrors to be used when setting up the skuba nodes, replacing the URL of the repositories defined in terraform. Used, for instance, to switch to development repositories or internal repositories when running in the CI pipeline.
```
non-essential clauses (like appositives) should be surrounded by commas

* plugin_dir: directory used for retrieving terraform plugins. If not set, plugins are installed using terraform [discovery mechanism](https://www.terraform.io/docs/extend/how-terraform-works.html#discovery)
* retries: maximum number of attempts to recover from failures during terraform provisioning 
* stack name: the unique name of the platform stack on the shared infrastructure, used as prefix by many resources such as networks, nodes, among others. If not specified, the `username` is used.
* tfdir: path to the terraform files. Testrunner must have writing permissions to this directory. Defaults to `skuba.srcpath/ci/infra`.
* tfvars: name of the terraform variables file to be used. Defaults to "terraform.tfvars.json.ci.example"

Example
```
terraform:
  stack_name="my-test-stack"
```

#### Openstack

> * openrc: path to the environment setup scrip

Reviewer 2: ```suggestion
* openrc: path to the environment setup script
```

Example:
```
openstack:
  openrc: "/home/myuser/my-openrc.sh"
```

#### VMware

* env_file: path to environment variables file

Example:
```
vmware: 
   env_file: "/path/to/env/file"
```

### Skuba

> The Skuba section defines the location and exection options for the `skuba` command. As `testrunner` can be used either from a local development or testing environment or a CI pipeline, the configuration allows to define the location of both the source and the binary. Please notice that the source location is used as default location for other configuration elements, such as terraform files, even if the `skuba` binary is specified. 

Reviewer 2: ```suggestion
The Skuba section defines the location and exectuion options for the `skuba` command. As `testrunner` can be used either from a local development or testing environment or a CI pipeline, the configuration allows to define the location of both the source and the binary. Please notice that the source location is used as default location for other configuration elements, such as terraform files, even if the `skuba` binary is specified. 
```

> The Skuba section defines the location and execution options for the `skuba` command. As `testrunner` can be used either from a local development or testing environment or a CI pipeline, the configuration allows to define the location of both the source and the binary. Please notice that the source location is used as default location for other configuration elements, such as terraform files, even if the `skuba` binary is specified. 

Reviewer 4: ```suggestion
The Skuba section defines the location and execution options for the `skuba` command. As `testrunner` can be used either from a CI pipeline or a local development or testing environment, the configuration allows defining the location of both the source and the binary. Please notice that the source location is used as the default location for other configuration elements, such as terraform files, even if the `skuba` binary is specified. 
```
Put single-part before the multi-part to clarify the "or", and a couple of minor grammatical tweaks

* binpath: path to skuba binary

> * srcpath: path to skuba source. Used to locate other resurces, like terraform configuration, and ssh keys.

Reviewer 2: ```suggestion
* srcpath: path to skuba source. Used to locate other resources, like terraform configuration, and ssh keys.
```

> * srcpath: path to skuba source. Used to locate other resources, like terraform configuration, and ssh keys.

Reviewer 4: ```suggestion
* srcpath: path to skuba source. Used to locate other resources, like terraform configuration and ssh keys.
```

* verbosity: verbosity level for skuba command execution

```
skuba:
  binpath: "/usr/bin/"
  srcpath: "/go/src/github.com/SUSE/skuba"
  verbosity: 5
```

### Log

Testrunner sends output to both a console and file logger handlers, configured using the following `log` variables:

* file: name of the file used to send a copy of the log with verbosity `DEBUG`. This file is located under the `workspace` directory.

> * leve: debug verbosity level to console. Can be any of `DEBUG`, `INFO`, `WARNING`, `ERROR`. Defaults to `INFO`.

Reviewer 2: ```suggestion
* level: debug verbosity level to console. Can be any of `DEBUG`, `INFO`, `WARNING`, `ERROR`. Defaults to `INFO`.
```

> * overwrite: boolean that indicates if the content of the log file must be overwrite (`True`) or log entries must be appended at the end of the file if it exists. Defaults to `False` (do not overwrite) 

Reviewer 2: ```suggestion
* overwrite: boolean that indicates if the content of the log file must be overwritten (`True`) or log entries must be appended at the end of the file if it exists. Defaults to `False` (do not overwrite) 
```

> * quiet: boolean that indicates if `testrunner` will send any output to console (`False`) or not will execute silently (`True`). Quiet mode is useful when `testrunner` is used as a library. Defaults to `False`.

Reviewer 4: Should it be indicated anywhere that the "True/False" booleans need to be in that specific case?  And, now that I type that, _are_ they actually case-sensitive in the YAML import? :D

Example:
```
log:
  level: DEBUG
```

### Test

> * no_destroy: boolean that indicates if provisioned resources should be deleted when test ends. Defaults to `False`

Reviewer 4: This double-negative is potentially confusing.  We should probably change that to "destroy" or similar and default it to `True`.

```
no_destroy: True  #keep resources after test ends
```

## Environment Setup

> This section details how to setup `testrunner` 

Reviewer 4: ```suggestion
This section details how to set up `testrunner` 
```

### Local setup

Copy `vars.yaml` to `/path/to/myvars.yaml`, set the variables according to your environment and needs, and use the `--vars /path/to/myvars.yaml` CLI argument when running `testrunner`.

#### Work Environment

> `testrunner` requires a working directory, which is specified in the `workspace` configuration parameter. The content of this directory can be erased or overwritten by `testrunner`. Be sure you crate a directory to be used as workspace which is not located under your local working copy of the `skuba` project.

Reviewer 2: ```suggestion
`testrunner` requires a working directory, which is specified in the `workspace` configuration parameter. The content of this directory can be erased or overwritten by `testrunner`. Be sure you create a directory to be used as workspace which is not located under your local working copy of the `skuba` project.
```

```
workspace: "/path/to/workspace"
username:  "my-user-test"
```

#### skuba and platform

Set the `skuba` and `terraform parameters depending on how you are testing `skuba`: 
* If testing from local source:
```
  skuba:
    srcpath: "/path/to/local/skuba/repo"

>     binpath: "path/to/go/bin/directory"

Reviewer 4: ```suggestion
    binpath: "/path/to/go/bin/directory"
```
Should this be an absolute path?  If relative is supported, that should be noted somewhere.

```

Be sure you don't specify the `terraform.tfdir` directory, so terraform configuration from the local `skuba` repo are used.

* If testing from installed package

```
  skuba:
    binpath: "/usr/bin/"

  terraform:
    tfdir: "/path/to/terraform/files"
```

> You use your `id_rsa` keys to connect to the cluster nodes, as the `shared_id` is not available.

Reviewer 3: this is configurable now

Author: Not yet. PR is not merged

Reviewer 4: I'll include an "update documentation" note in #593; there are a couple of places where the README will need tweaked after that change is applied.

Author: That PR is not yet merged and actually I asked for a change, so let's update the documentation as part of that PR.

export VSPHERE_ALLOW_UNVERIFIED_SSL="true"
```

2. Set the path to the VMware environment file in the testrunner's vars file:

```
vmware:

>   env_file: "=/path/to/vmware-env.sh"

Reviewer 4: ```suggestion
  env_file: "/path/to/vmware-env.sh"
```
I assume the leading `=` is a typo, since it's not in the environment variable below?

```
or as an environment variable: `export VMWARE_ENV_FILE=/path/to/vmware-env.sh`

3. Be sure to use the `-p` or `--platform` argument when invoking `testrunner` and set it to `vmware`, otherwise `openstack` is used. 

### Jenkins Setup

> In your Jenkins file, you need to set up environment variables. Then These environment variables will replace variables in the yaml file.

Reviewer 2: ```suggestion
In your Jenkins file, you need to set up environment variables. Then these environment variables will replace variables in the yaml file.
```

> In your Jenkins file, you need to set up environment variables which will replace options in the yaml file. This result more convenient than having to edit the yaml file in the CI pipeline. 

Reviewer 4: ```suggestion
In your Jenkins file, you need to set up environment variables which will replace options in the yaml file. This is more convenient than having to edit the yaml file in the CI pipeline. 
```

#### Work environment

> As default, Jenkins has `WORKSPACE` environment variable so that `workspace` will be replaced automatically.

Reviewer 4: ```suggestion
By default, Jenkins has a `WORKSPACE` environment variable so that `workspace` will be replaced automatically.
```

#### Skuba

> Jenkins checkouts the `skuba` repository under the `workspace` directory and generates the binaries also under the `workspace`, which are the default locations. Therefore, there is no need to specify any location:

Reviewer 2: ```suggestion
Jenkins checks out the `skuba` repository under the `workspace` directory and generates the binaries also under the `workspace`, which are the default locations. Therefore, there is no need to specify any location:
```

  -v, --verbose         show all output
```

## Examples 

### Create K8s Cluster

1. Deploy nodes to openstack

> ```./testrunner provision```

Author: I think it is unnecessary as the README is already in that location

> ```./testrunner provision```

Reviewer 4: The README is on the Internet at GitHub when I read it. ;)  But unless we move testrunner to a different directory, I think people will probably figure it out.
